### ISSUE TITLE
Incorrect calculation of 2nd derivative of a determinant of a matrix

### ISSUE THREAD

**Issue by coreyjadams** [tensorflow/tensorflow#55230]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Mac OS Monterey.  Also seen on Linux.
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.8.0-rc1-32-g3f878cff5b6 2.8.0 (also seen in 2.6.X)
- Python version: 3.8.6
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A, but also seen in cuda11.4
- GPU model and memory: CPU, but also seen in A100 40GB/80GB

**Describe the current behavior**

The calculation of the second derivative of a determinant is incorrect. 

It is a little challenging to describe the full issue in markdown.  I've created a standalone notebook to reproduce the issue.  I compare the TF gradients through a determinant against both finite differences and a custom (albeit slow) determinant implementation.  The TF gradients for the determinant of a matrix, let's call them G, agree with both finite differences and the custom op.  The second derivative, dG/dx, aka the Hessian of the determinant of a matrix, is badly incorrect.  Using autodifferentiation on the custom determinant operation twice, however, agrees well with finite differences methods.

Full reproducer available in this notebook (standalone) here: https://github.com/Nuclear-Physics-with-Machine-Learning/MLQM/blob/spin/examples/2nd%20Derivative%20of%20Determinant%20of%20a%20matrix.ipynb

**Describe the expected behavior**

The calculation of the second derivative of a determinant should be correct.

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no):  If we can track down the bug, and it's something i can fix, I could.  I haven't modified yet.
- Briefly describe your candidate solution(if contributing):  I believe the operations registered for the backprop through a determinant of a matrix must be missing something, and this could be fixed.

**Standalone code to reproduce the issue**

https://github.com/Nuclear-Physics-with-Machine-Learning/MLQM/blob/spin/examples/2nd%20Derivative%20of%20Determinant%20of%20a%20matrix.ipynb

**Comment by tilakrayal** [comment]
@gadagashwini ,
I was able to reproduce the issue in tf [v2.8](https://colab.research.google.com/gist/tilakrayal/0bd5fcb3ec4eebee43faf38bd81a5b13/55230.ipynb), [v2.7](https://colab.research.google.com/gist/tilakrayal/5683fc7edd39e62db6d23bf70d7e07a7/2-7-55230.ipynb) and [nightly](https://colab.research.google.com/gist/tilakrayal/cbe58602a2e27c83d33aa0282bd24376/nightly55230.ipynb).Please find the gist attached.

**Comment by coreyjadams** [comment]
Thanks for reproducing it.  I wasn't able to track down where this originates in Tensorflow.  I found the function that computes the derivative of det(A), but I don't know enough about tensorflow or the linalg package to try debugging this.  How can I help?

**Comment by coreyjadams** [comment]
Hi there, any thoughts on this?  Anything I can do to help?

**Comment by coreyjadams** [comment]
Another ping ... any news?

**Comment by tomytsai1** [comment]
Hi @coreyjadams ,
We seemed to find the root cause, but are working on the solution.

The problem is from [linalg_grad._MatrixDeterminantGrad](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/linalg_grad.py#L376) where the matrix inverse function was invoked with `adjoint = True` flag, but when taking the second derivative the adjoint flag was not correctly considered, so the computation was wrong.  Hopefully the fix is not hard and will be out soon.

**Comment by tomytsai1** [comment]
The fix has been [committed](https://github.com/tensorflow/tensorflow/commit/16e3b5862e52876863529ab481a71f81b44a1df7) yesterday. Let us know if you still see any issues.
